Message traits

Xanewok opened the pull request:
Attempt at #19. I took some sweet time with this, sorry for doing it so late.

Currently LSP spec provides new `CompletionParams` and `CompletionContext` types to be used with Completion Request, so I used LSP commit in the comment and used [`TextDocumentPositionParams`](https://github.com/gluon-lang/languageserver-types/compare/master...Xanewok:message-traits?expand=1#diff-b4aea3e418ccdb71239b96952d9cddb6R1643) as param type instead.

@Marwes does this align with what you wanted in #19?

Marwes commented:
Yep! This is pretty much what I wanted in #19. I actually did some further experimentation with this idea a few weeks ago but it never went anywhere since I was never completely happy with the result. Basically I wanted a way to also generate a trait with handler methods as well such as [RustLSP uses](https://github.com/gluon-lang/languageserver-types/issues/19#issuecomment-312850272) but where it was fully customisable with respect to method return types. It turned out to not quite work so my efforts fizzled out at that point without me actually implementing this part even...

I think one idea from it can be salvaged to work with this PR. I thought that it might be better to describe the method/notification in terms of their actual name in the spec. So instead of `PublishDiagnostics` it would be `request!("textDocument/publishDiagnostics")`. It is perhaps a bit verbose but it might be easier to grep. Could try and implement it for real if it sounds useful, I am more or less undecided at this point.

Marwes commented on src/lib.rs:
```diff
@@ -49,6 +49,19 @@ pub enum NumberOrString {
+pub trait Notification {
+    type Params;
+    const METHOD: &'static str;
+}
```
I wonder, would it be better to move the traits and their implementations to submodules so they are not confused with the data holding types? `mod rpc`, `mod notification + mod request`?

Xanewok replied:
Yeah, actually the thought crossed my mind as well. I'll try to separate it into `mod notification + mod request`.

Xanewok replied:
@Marwes should I move there helper union return values that aren't defined by the protocol, but needed to encode the result type? e.g. https://github.com/gluon-lang/languageserver-types/pull/29/files#diff-b4aea3e418ccdb71239b96952d9cddb6R2000

Marwes replied:
I could see either place being good. Maybe an argument for keeping them here is that it makes the submodules only hold types that you aren't meant to instantiate (since they don't hold data).

Marwes replied:
Keeping them here means the result type you linked to.

Marwes commented on src/lib.rs:
```diff
@@ -57,6 +70,14 @@ pub enum NumberOrString {
 /// It can not be left open / hanging. This is in line with the JSON RPC protocol that requires
 /// that every request sends a response back. In addition it allows for returning partial results on cancel.
+#[derive(Debug)]
+pub struct Cancel;
```
Since these types aren't meant to be used as values we could have them as uninhabited enums maybe?

```rust
pub enum Cancel { }
```

That way they can only be used in as types  and not values. Or `pub struct Cancel(());` so they can't be instantiated.

Xanewok commented:
Rebased and force-pushed. I toyed with a simple `request!` macro generating type and impl, to not repeat that much, but I couldn't get doc comments to work with the macro-generated type, so I dropped it for now.

Marwes commented:
Looks good now I think!

For doc comments/attributes in macros this should match on them, in case you need it in the future https://github.com/Marwes/combine/blob/0dba46ca7a8682e1a048dee7fc9b63d6a588ea21/src/lib.rs#L295-L301

Marwes commented:
Publishes as 0.17.0

I will probably try and port gluon's language server to this over the weekend :)

Xanewok commented:
Cool, thanks! RLS will probably also benefit from this as well :smile:
